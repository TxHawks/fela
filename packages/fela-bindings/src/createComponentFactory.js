/* @flow */
import {
  hoistStatics,
  extractPassThroughProps,
  extractUsedProps,
  resolvePassThrough,
  resolveUsedProps
} from 'fela-utils'
import { combineRules } from 'fela'

export default function createComponentFactory(
  createElement: Function,
  withTheme: Function,
  contextTypes?: Object,
  withProxy: boolean = false
): Function {
  return function createComponent(
    rule: Function,
    type: any = 'div',
    passThroughProps: Array<string> | Function = []
  ): Function {
    const displayName = rule.name ? rule.name : 'FelaComponent'

    const FelaComponent = (
      {
        children,
        theme,
        _felaRule,
        extend,
        innerRef,
        id,
        style,
        as,
        className,
        passThrough = [],
        ...otherProps
      },
      { renderer }
    ) => {
      if (!renderer) {
        throw new Error(
          "createComponent() can't render styles without the renderer in the context. Missing react-fela's <Provider /> at the app root?"
        )
      }

      const usedProps = withProxy ? extractUsedProps(rule, theme) : {}
      const composedRule = _felaRule ? combineRules(rule, _felaRule) : rule
      const combinedRule = extend
        ? combineRules(composedRule, () => extend)
        : composedRule

      // improve developer experience with monolithic renderer
      if (process.env.NODE_ENV !== 'production' && renderer.prettySelectors) {
        const componentName =
          typeof type === 'string' ? type : type.displayName || type.name || ''

        combinedRule.selectorPrefix = `${displayName}_${componentName}__`
      }
      // compose passThrough props from arrays or functions
      const resolvedPassThrough = [
        ...resolvePassThrough(passThroughProps, otherProps),
        ...resolvePassThrough(passThrough, otherProps),
        ...(withProxy ? resolveUsedProps(usedProps, otherProps) : [])
      ]

      const ruleProps = {
        ...otherProps,
        theme
      }

      // if the component renders into another Fela component
      // we pass down the combinedRule as well as both
      if (type._isFelaComponent) {
        return createElement(
          type,
          {
            _felaRule: combinedRule,
            passThrough: resolvedPassThrough,
            ...ruleProps
          },
          children
        )
      }

      const componentProps = extractPassThroughProps(
        resolvedPassThrough,
        otherProps
      )

      // fela-native support
      if (renderer.isNativeRenderer) {
        const felaStyle = renderer.renderRule(combinedRule, ruleProps)
        componentProps.style = style ? [style, felaStyle] : felaStyle
      } else {
        if (style) {
          componentProps.style = style
        }

        const cls = className ? `${className} ` : ''
        componentProps.className =
          cls + renderer.renderRule(combinedRule, ruleProps)
      }

      if (id) {
        componentProps.id = id
      }

      if (innerRef) {
        componentProps.ref = innerRef
      }

      const customType = as || type
      return createElement(customType, componentProps, children)
    }

    if (contextTypes) {
      FelaComponent.contextTypes = contextTypes
    }

    // use the rule name as display name to better debug with react inspector
    FelaComponent.displayName = displayName
    FelaComponent._isFelaComponent = true

    const themedComponent = withTheme(FelaComponent)
    return hoistStatics(themedComponent, type)
  }
}
