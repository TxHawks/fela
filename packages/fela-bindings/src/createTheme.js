/* @flow */
import arrayEach from 'fast-loops/lib/arrayEach'

import type { Theme } from '../../../flowtypes/Theme'

export default function createTheme(
  themeProperties: Object = {},
  previousTheme: Theme
): Theme {
  const theme: Theme = {
    listeners: [],
    properties: themeProperties,

    update(newProperties: Object): void {
      theme.properties = newProperties
      theme._emitChange()
    },

    get(): Object {
      return {
        ...theme.previousProperties,
        ...theme.properties,
      }
    },

    subscribe(listener: Function): Function {
      theme.listeners.push(listener)

      return () => theme.listeners.splice(theme.listeners.indexOf(listener), 1)
    },

    _emitChange(): void {
      const properties = theme.get()

      arrayEach(theme.listeners, listener => listener(properties))
    },
  }

  if (previousTheme) {
    theme.previousProperties = previousTheme.get()
    previousTheme.subscribe(properties => {
      theme.previousProperties = properties
      theme._emitChange()
    })
  } else {
    theme.previousProperties = {}
  }

  return theme
}
