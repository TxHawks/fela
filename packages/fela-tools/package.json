{
  "name": "fela-tools",
  "version": "5.0.8",
  "description": "Useful tools for working with Fela",
  "main": "lib/index.js",
  "module": "es/index.js",
  "jsnext:main": "es/index.js",
  "files": [
    "LICENSE",
    "README.md",
    "lib/**",
    "es/**"
  ],
  "repository": "https://github.com/rofrischmann/fela/",
  "keywords": [
    "fela",
    "fela-tool",
    "cssinjs"
  ],
  "author": "Robin Frischmann",
  "license": "MIT",
  "dependencies": {
    "css-in-js-utils": "^2.0.0",
    "fela-utils": "^7.0.0"
  },
  "devDependencies": {
    "fela": "^6.0.0"
  }
}
