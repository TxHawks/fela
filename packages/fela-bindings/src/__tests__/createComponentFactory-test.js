import React, { createElement, Component as BaseComponent } from 'react'
import PropTypes from 'prop-types'
import { mount } from 'enzyme'
import toJson from 'enzyme-to-json'
import { html as beautify } from 'js-beautify'

import { renderToString } from 'fela-tools'
import { createRenderer } from 'fela'
import monolithic from 'fela-monolithic'

import createComponentFactory from '../createComponentFactory'
import withThemeFactory from '../withThemeFactory'
import createTheme from '../createTheme'

const withTheme = withThemeFactory(BaseComponent, createElement, {
  theme: PropTypes.object
})

const createComponent = createComponentFactory(createElement, withTheme, {
  renderer: PropTypes.object
})

const createComponentWithProxy = createComponentFactory(
  createElement,
  withTheme,
  {
    renderer: PropTypes.object
  },
  true
)

describe('Creating Components from Fela rules', () => {
  it('should return a Component', () => {
    const rule = props => ({
      color: props.color,
      fontSize: 16
    })

    const component = createComponent(rule)

    expect(component).toBeInstanceOf(Function)
  })

  it('should render fela rules depending on the passed props', () => {
    const rule = props => ({
      color: props.color,
      fontSize: 16
    })

    const renderer = createRenderer()
    const Component = createComponent(rule)

    const wrapper = mount(<Component color="black" />, {
      context: {
        renderer
      }
    })

    expect([
      beautify(`<style>${renderToString(renderer)}</style>`),
      toJson(wrapper)
    ]).toMatchSnapshot()
  })

  it('should include defaultProps if provided', () => {
    const rule = props => ({
      color: props.color,
      fontSize: 16
    })

    const Comp = ({ color, className }) => (
      <div className={className}>{color}</div>
    )

    Comp.defaultProps = {
      color: 'red'
    }

    const Component = createComponent(rule, Comp)
    const renderer = createRenderer()

    const wrapper = mount(<Component />, {
      context: {
        renderer
      }
    })

    expect([
      beautify(`<style>${renderToString(renderer)}</style>`),
      toJson(wrapper)
    ]).toMatchSnapshot()
  })

  it('should use the theme for static rendering by default', () => {
    const rule = props => ({
      color: props.theme.color,
      fontSize: 16
    })

    const Component = createComponent(rule)
    const renderer = createRenderer()
    const theme = createTheme({
      color: 'red'
    })

    const wrapper = mount(<Component />, {
      context: {
        renderer,
        theme
      }
    })

    expect([
      beautify(`<style>${renderToString(renderer)}</style>`),
      toJson(wrapper)
    ]).toMatchSnapshot()
  })

  it('should allow nested theme objects', () => {
    const rule = props => ({
      color: props.theme.header.color,
      fontSize: 16
    })

    const Component = createComponent(rule)
    const renderer = createRenderer()
    const theme = createTheme({
      header: {
        color: 'black'
      }
    })

    const wrapper = mount(<Component />, {
      context: {
        renderer,
        theme
      }
    })

    expect([
      beautify(`<style>${renderToString(renderer)}</style>`),
      toJson(wrapper)
    ]).toMatchSnapshot()
  })

  it('should not pass props to the element', () => {
    const rule = props => ({
      color: props.color,
      fontSize: 16
    })

    const Component = createComponent(rule, 'div', ['onClick'])
    const renderer = createRenderer()

    const wrapper = mount(<Component onClick={false} onHover color />, {
      context: {
        renderer
      }
    })

    expect([
      beautify(`<style>${renderToString(renderer)}</style>`),
      toJson(wrapper)
    ]).toMatchSnapshot()
  })

  it('should pass all props to the element', () => {
    const rule = props => ({
      color: props.color,
      fontSize: 16
    })

    const Component = createComponent(rule, 'div', Object.keys)
    const renderer = createRenderer()

    const wrapper = mount(<Component onClick={false} data-foo />, {
      context: {
        renderer
      }
    })

    expect([
      beautify(`<style>${renderToString(renderer)}</style>`),
      toJson(wrapper)
    ]).toMatchSnapshot()
  })

  it('should only use passed props to render Fela rules', () => {
    const rule = props => ({
      color: props['data-foo'] && props.color,
      fontSize: '16px'
    })

    const Component = createComponent(rule, 'div', ['data-foo'])
    const renderer = createRenderer()

    const wrapper = mount(<Component data-foo color="black" />, {
      context: {
        renderer
      }
    })

    expect([
      beautify(`<style>${renderToString(renderer)}</style>`),
      toJson(wrapper)
    ]).toMatchSnapshot()
  })

  it('should compose styles', () => {
    const rule = () => ({
      color: 'blue',
      fontSize: '16px'
    })

    const anotherRule = () => ({
      color: 'red',
      lineHeight: 1.2
    })

    const Comp = createComponent(rule)
    const ComposedComp = createComponent(anotherRule, Comp)
    const renderer = createRenderer()

    const wrapper = mount(<ComposedComp />, {
      context: {
        renderer
      }
    })

    expect([
      beautify(`<style>${renderToString(renderer)}</style>`),
      toJson(wrapper)
    ]).toMatchSnapshot()
  })

  it('should extend the rule properties', () => {
    const rule = () => ({
      color: 'blue',
      fontSize: '16px'
    })

    const Comp = createComponent(rule)
    const renderer = createRenderer()

    const bgColor = 'red'

    const wrapper = mount(
      <Comp
        extend={{
          fontSize: '14px',
          backgroundColor: bgColor
        }}
      />,
      {
        context: {
          renderer
        }
      }
    )

    expect([
      beautify(`<style>${renderToString(renderer)}</style>`),
      toJson(wrapper)
    ]).toMatchSnapshot()
  })

  it('should compose passThrough props', () => {
    const Component = createComponent(() => ({}), 'div', Object.keys)
    const ComposedComponent = createComponent(() => ({}), Component, [
      'onClick'
    ])

    const renderer = createRenderer()

    const wrapper = mount(
      <ComposedComponent onClick={() => true} color="red" />,
      {
        context: {
          renderer
        }
      }
    )

    expect([
      beautify(`<style>${renderToString(renderer)}</style>`),
      toJson(wrapper)
    ]).toMatchSnapshot()
  })

  it('should only use the rule name as displayName', () => {
    const Button = () => ({
      color: 'red',
      fontSize: 16
    })
    const component = createComponent(Button)

    expect(component.displayName).toEqual('Button')
  })

  it('should use a dev-friendly className with monolithic renderer', () => {
    const Button = () => ({
      fontSize: 16
    })

    const Component = createComponent(Button)

    const renderer = createRenderer({
      enhancers: [
        monolithic({
          prettySelectors: true
        })
      ]
    })

    const wrapper = mount(<Component />, {
      context: {
        renderer
      }
    })

    expect([
      beautify(`<style>${renderToString(renderer)}</style>`),
      toJson(wrapper)
    ]).toMatchSnapshot()
  })

  it('should use a dev-friendly className and the selectorPrefix', () => {
    const Button = () => ({
      fontSize: 16
    })

    const Component = createComponent(Button)

    const renderer = createRenderer({
      enhancers: [
        monolithic({
          prettySelectors: true
        })
      ],
      selectorPrefix: 'Fela-'
    })

    const wrapper = mount(<Component />, {
      context: {
        renderer
      }
    })

    expect([
      beautify(`<style>${renderToString(renderer)}</style>`),
      toJson(wrapper)
    ]).toMatchSnapshot()
  })

  it('should only use the rule name as displayName', () => {
    const Button = () => ({
      color: 'red',
      fontSize: 16
    })
    const Component = createComponent(Button)
    const renderer = createRenderer()

    const wrapper = mount(<Component as="button" />, {
      context: {
        renderer
      }
    })

    expect([
      beautify(`<style>${renderToString(renderer)}</style>`),
      toJson(wrapper)
    ]).toMatchSnapshot()
  })
})

describe('Creating Components with a Proxy for props from Fela rules', () => {
  it('should return a Component', () => {
    const rule = props => ({
      color: props.color,
      fontSize: 16
    })
    const component = createComponentWithProxy(rule)

    expect(component).toBeInstanceOf(Function)
  })

  it('should render fela rules depending on the passed props', () => {
    const rule = props => ({
      color: props.color,
      fontSize: 16
    })

    const Component = createComponentWithProxy(rule)

    const renderer = createRenderer()

    const wrapper = mount(<Component color="black" />, {
      context: {
        renderer
      }
    })

    expect([
      beautify(`<style>${renderToString(renderer)}</style>`),
      toJson(wrapper)
    ]).toMatchSnapshot()
  })

  it('should use the theme for static rendering by default', () => {
    const rule = props => ({
      color: props.theme.color,
      fontSize: 16
    })
    const Component = createComponentWithProxy(rule)
    const renderer = createRenderer()
    const theme = createTheme({
      color: 'black'
    })

    const wrapper = mount(<Component />, {
      context: {
        renderer,
        theme
      }
    })

    expect([
      beautify(`<style>${renderToString(renderer)}</style>`),
      toJson(wrapper)
    ]).toMatchSnapshot()
  })

  it('should allow nested theme objects', () => {
    const rule = props => ({
      color: props.theme.header.color,
      fontSize: 16
    })
    const Component = createComponentWithProxy(rule)
    const renderer = createRenderer()
    const theme = createTheme({
      header: {
        color: 'black'
      }
    })

    const wrapper = mount(<Component />, {
      context: {
        renderer,
        theme
      }
    })

    expect([
      beautify(`<style>${renderToString(renderer)}</style>`),
      toJson(wrapper)
    ]).toMatchSnapshot()
  })

  it('should not pass props used in rules to the element', () => {
    const rule = props => ({
      color: props.color,
      fontSize: 16
    })
    const Component = createComponentWithProxy(rule, 'div')

    const renderer = createRenderer()

    const wrapper = mount(<Component onClick={false} data-foo color />, {
      context: {
        renderer
      }
    })

    expect([
      beautify(`<style>${renderToString(renderer)}</style>`),
      toJson(wrapper)
    ]).toMatchSnapshot()
  })

  it('should pass props used in rules specified in passThroughProps to the element', () => {
    const rule = props => ({
      color: props.color,
      fontSize: 16
    })
    const Component = createComponentWithProxy(rule, 'div', ['color'])

    const renderer = createRenderer()

    const wrapper = mount(<Component onClick={false} data-foo color />, {
      context: {
        renderer
      }
    })

    expect([
      beautify(`<style>${renderToString(renderer)}</style>`),
      toJson(wrapper)
    ]).toMatchSnapshot()
  })

  it('should pass all props to the element', () => {
    const rule = props => ({
      color: props.color,
      fontSize: 16
    })

    const Component = createComponentWithProxy(rule, 'div', Object.keys)

    const renderer = createRenderer()

    const wrapper = mount(<Component onClick={false} data-foo />, {
      context: {
        renderer
      }
    })

    expect([
      beautify(`<style>${renderToString(renderer)}</style>`),
      toJson(wrapper)
    ]).toMatchSnapshot()
  })

  it('should only use passed props to render Fela rules', () => {
    const rule = props => ({
      color: props['data-foo'] && props.color,
      fontSize: '16px'
    })

    const Component = createComponentWithProxy(rule, 'div', ['data-foo'])
    const renderer = createRenderer()

    const wrapper = mount(<Component data-foo color="black" />, {
      context: {
        renderer
      }
    })

    expect([
      beautify(`<style>${renderToString(renderer)}</style>`),
      toJson(wrapper)
    ]).toMatchSnapshot()
  })

  it('should compose styles', () => {
    const rule = () => ({
      color: 'blue',
      fontSize: '16px'
    })

    const anotherRule = () => ({
      color: 'red',
      lineHeight: 1.2
    })

    const Comp = createComponentWithProxy(rule)
    const ComposedComp = createComponentWithProxy(anotherRule, Comp)

    const renderer = createRenderer()

    const wrapper = mount(<ComposedComp />, {
      context: {
        renderer
      }
    })

    expect([
      beautify(`<style>${renderToString(renderer)}</style>`),
      toJson(wrapper)
    ]).toMatchSnapshot()
  })

  it('should compose passThrough props', () => {
    const Comp = createComponentWithProxy(() => ({}), 'div', Object.keys)
    const ComposedComp = createComponentWithProxy(() => ({}), Comp, ['onClick'])

    const renderer = createRenderer()

    const onClick = () => true
    const wrapper = mount(<ComposedComp onClick={onClick} color="red" />, {
      context: {
        renderer
      }
    })

    expect([
      beautify(`<style>${renderToString(renderer)}</style>`),
      toJson(wrapper)
    ]).toMatchSnapshot()
  })

  it('should only use the rule name as displayName', () => {
    const Button = () => ({
      color: 'red',
      fontSize: 16
    })
    const component = createComponentWithProxy(Button)

    expect(component.displayName).toEqual('Button')
  })

  it('should use a dev-friendly className with monolithic renderer', () => {
    const Button = () => ({
      fontSize: 16
    })

    const Component = createComponentWithProxy(Button)

    const renderer = createRenderer({
      enhancers: [
        monolithic({
          prettySelectors: true
        })
      ]
    })

    const wrapper = mount(<Component />, {
      context: {
        renderer
      }
    })

    expect([
      beautify(`<style>${renderToString(renderer)}</style>`),
      toJson(wrapper)
    ]).toMatchSnapshot()
  })

  it('should use a dev-friendly className and the selectorPrefix', () => {
    const Button = () => ({
      fontSize: 16
    })

    const Component = createComponentWithProxy(Button)

    const renderer = createRenderer({
      enhancers: [
        monolithic({
          prettySelectors: true
        })
      ],
      selectorPrefix: 'Fela-'
    })

    const wrapper = mount(<Component />, {
      context: {
        renderer
      }
    })

    expect([
      beautify(`<style>${renderToString(renderer)}</style>`),
      toJson(wrapper)
    ]).toMatchSnapshot()
  })

  it('should only use the rule name as displayName', () => {
    const Button = () => ({
      color: 'red',
      fontSize: 16
    })
    const Component = createComponentWithProxy(Button)
    const renderer = createRenderer()

    const wrapper = mount(<Component as="button" />, {
      context: {
        renderer
      }
    })

    expect([
      beautify(`<style>${renderToString(renderer)}</style>`),
      toJson(wrapper)
    ]).toMatchSnapshot()
  })
})

it('should pass props except innerRef', () => {
  const rule = props => ({
    color: props.color,
    fontSize: '16px'
  })
  const Component = createComponentWithProxy(rule, 'div')

  const renderer = createRenderer()

  const wrapper = mount(<Component color="black" innerRef={() => 'test'} />, {
    context: {
      renderer
    }
  })

  expect([
    beautify(`<style>${renderToString(renderer)}</style>`),
    toJson(wrapper)
  ]).toMatchSnapshot()
})
