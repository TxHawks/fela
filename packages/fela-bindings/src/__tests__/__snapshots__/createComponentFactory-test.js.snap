// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Creating Components from Fela rules should allow nested theme objects 1`] = `
Array [
  "<style>
    .a {
        color: black
    }

    .b {
        font-size: 16
    }
</style>",
  <rule>
    <rule
        _felaTheme={
            Object {
                "header": Object {
                  "color": "black",
                },
              }
        }
    >
        <div
            className="a b"
        />
    </rule>
</rule>,
]
`;

exports[`Creating Components from Fela rules should compose passThrough props 1`] = `
Array [
  "<style></style>",
  <FelaComponent
    color="red"
    onClick={[Function]}
>
    <FelaComponent
        _felaTheme={Object {}}
        color="red"
        onClick={[Function]}
    >
        <FelaComponent
            _felaRule={[Function]}
            color="red"
            onClick={[Function]}
            passThrough={
                Array [
                    "onClick",
                  ]
            }
        >
            <FelaComponent
                _felaRule={[Function]}
                _felaTheme={Object {}}
                color="red"
                onClick={[Function]}
                passThrough={
                    Array [
                        "onClick",
                      ]
                }
            >
                <div
                    className=""
                    color="red"
                    onClick={[Function]}
                />
            </FelaComponent>
        </FelaComponent>
    </FelaComponent>
</FelaComponent>,
]
`;

exports[`Creating Components from Fela rules should compose styles 1`] = `
Array [
  "<style>
    .a {
        color: red
    }

    .b {
        font-size: 16px
    }

    .c {
        line-height: 1.2
    }
</style>",
  <anotherRule>
    <anotherRule
        _felaTheme={Object {}}
    >
        <rule
            _felaRule={[Function]}
            passThrough={Array []}
        >
            <rule
                _felaRule={[Function]}
                _felaTheme={Object {}}
                passThrough={Array []}
            >
                <div
                    className="a b c"
                />
            </rule>
        </rule>
    </anotherRule>
</anotherRule>,
]
`;

exports[`Creating Components from Fela rules should extend the rule properties with a function 1`] = `
Array [
  "<style>
    .a {
        color: blue
    }

    .b {
        font-size: 14px
    }

    .c {
        background-color: red
    }
</style>",
  <rule
    bgColor="red"
    extend={[Function]}
>
    <rule
        _felaTheme={Object {}}
        bgColor="red"
        extend={[Function]}
    >
        <div
            className="a b c"
        />
    </rule>
</rule>,
]
`;

exports[`Creating Components from Fela rules should extend the rule properties with an object 1`] = `
Array [
  "<style>
    .a {
        color: blue
    }

    .b {
        font-size: 14px
    }

    .c {
        background-color: red
    }
</style>",
  <rule
    extend={
        Object {
            "backgroundColor": "red",
            "fontSize": "14px",
          }
    }
>
    <rule
        _felaTheme={Object {}}
        extend={
            Object {
                "backgroundColor": "red",
                "fontSize": "14px",
              }
        }
    >
        <div
            className="a b c"
        />
    </rule>
</rule>,
]
`;

exports[`Creating Components from Fela rules should include defaultProps if provided 1`] = `
Array [
  "<style>
    .a {
        color: red
    }

    .b {
        font-size: 16
    }
</style>",
  <rule
    color="red"
>
    <rule
        _felaTheme={Object {}}
        color="red"
    >
        <Comp
            className="a b"
            color="red"
        >
            <div
                className="a b"
            >
                red
            </div>
        </Comp>
    </rule>
</rule>,
]
`;

exports[`Creating Components from Fela rules should not pass props to the element 1`] = `
Array [
  "<style>
    .a {
        color: true
    }

    .b {
        font-size: 16
    }
</style>",
  <rule
    color={true}
    onClick={false}
    onHover={true}
>
    <rule
        _felaTheme={Object {}}
        color={true}
        onClick={false}
        onHover={true}
    >
        <div
            className="a b"
            onClick={false}
        />
    </rule>
</rule>,
]
`;

exports[`Creating Components from Fela rules should only use passed props to render Fela rules 1`] = `
Array [
  "<style>
    .a {
        color: black
    }

    .b {
        font-size: 16px
    }
</style>",
  <rule
    color="black"
    data-foo={true}
>
    <rule
        _felaTheme={Object {}}
        color="black"
        data-foo={true}
    >
        <div
            className="a b"
            data-foo={true}
        />
    </rule>
</rule>,
]
`;

exports[`Creating Components from Fela rules should only use the rule name as displayName 1`] = `
Array [
  "<style>
    .a {
        color: red
    }

    .b {
        font-size: 16
    }
</style>",
  <Button
    as="button"
>
    <Button
        _felaTheme={Object {}}
        as="button"
    >
        <button
            className="a b"
        />
    </Button>
</Button>,
]
`;

exports[`Creating Components from Fela rules should pass all props to the element 1`] = `
Array [
  "<style>
    .a {
        font-size: 16
    }
</style>",
  <rule
    data-foo={true}
    onClick={false}
>
    <rule
        _felaTheme={Object {}}
        data-foo={true}
        onClick={false}
    >
        <div
            className="a"
            data-foo={true}
            onClick={false}
        />
    </rule>
</rule>,
]
`;

exports[`Creating Components from Fela rules should pass extended special props to the component 1`] = `
Array [
  "<style>
    .a {
        color: red
    }

    .b {
        font-size: 16
    }
</style>",
  <rule
    color="blue"
    testProp={
        Object {
            "tabIndex": "-1",
          }
    }
>
    <rule
        _felaTheme={Object {}}
        color="blue"
        testProp={
            Object {
                "tabIndex": "-1",
              }
        }
    >
        <UnderlyingComp
            className="a b"
            testProp={
                Object {
                    "tabIndex": "-1",
                  }
            }
        >
            <div>
                testProp exists
            </div>
        </UnderlyingComp>
    </rule>
</rule>,
  <rule
    color="blue"
>
    <rule
        _felaTheme={Object {}}
        color="blue"
    >
        <UnderlyingComp
            className="a b"
        >
            <div>
                testProp does not exist
            </div>
        </UnderlyingComp>
    </rule>
</rule>,
]
`;

exports[`Creating Components from Fela rules should pass special props to the component 1`] = `
Array [
  "<style>
    .a {
        color: blue
    }

    .b {
        font-size: 16
    }

    .c {
        color: red
    }
</style>",
  <rule
    as="i"
    color="blue"
>
    <rule
        _felaTheme={Object {}}
        as="i"
        color="blue"
    >
        <i
            className="a b"
        />
    </rule>
</rule>,
  <rule
    color="blue"
>
    <rule
        _felaTheme={Object {}}
        color="blue"
    >
        <div
            className="c b"
        />
    </rule>
</rule>,
]
`;

exports[`Creating Components from Fela rules should pass style, as, id, className and innerRef to composed components 1`] = `
Array [
  "<style>
    .a {
        color: green
    }

    .b {
        font-size: 16px
    }

    .c {
        line-height: 1.2
    }
</style>",
  <anotherRule
    as="i"
    color="green"
>
    <anotherRule
        _felaTheme={Object {}}
        as="i"
        color="green"
    >
        <rule
            _felaRule={[Function]}
            as="i"
            color="green"
            passThrough={Array []}
        >
            <rule
                _felaRule={[Function]}
                _felaTheme={Object {}}
                as="i"
                color="green"
                passThrough={Array []}
            >
                <i
                    className="a b c"
                />
            </rule>
        </rule>
    </anotherRule>
</anotherRule>,
]
`;

exports[`Creating Components from Fela rules should render fela rules depending on the passed props 1`] = `
Array [
  "<style>
    .a {
        color: black
    }

    .b {
        font-size: 16
    }
</style>",
  <rule
    color="black"
>
    <rule
        _felaTheme={Object {}}
        color="black"
    >
        <div
            className="a b"
        />
    </rule>
</rule>,
]
`;

exports[`Creating Components from Fela rules should use a dev-friendly className and the selectorPrefix 1`] = `
Array [
  "<style>
    .Fela-Button_div__abrv9k {
        font-size: 16
    }
</style>",
  <Button>
    <Button
        _felaTheme={Object {}}
    >
        <div
            className="Fela-Button_div__abrv9k"
        />
    </Button>
</Button>,
]
`;

exports[`Creating Components from Fela rules should use a dev-friendly className with monolithic renderer 1`] = `
Array [
  "<style>
    .Button_div__abrv9k {
        font-size: 16
    }
</style>",
  <Button>
    <Button
        _felaTheme={Object {}}
    >
        <div
            className="Button_div__abrv9k"
        />
    </Button>
</Button>,
]
`;

exports[`Creating Components from Fela rules should use the theme for static rendering by default 1`] = `
Array [
  "<style>
    .a {
        color: red
    }

    .b {
        font-size: 16
    }
</style>",
  <rule>
    <rule
        _felaTheme={
            Object {
                "color": "red",
              }
        }
    >
        <div
            className="a b"
        />
    </rule>
</rule>,
]
`;

exports[`Creating Components with a Proxy for props from Fela rules should allow nested theme objects 1`] = `
Array [
  "<style>
    .a {
        color: black
    }

    .b {
        font-size: 16
    }
</style>",
  <rule>
    <rule
        _felaTheme={
            Object {
                "header": Object {
                  "color": "black",
                },
              }
        }
    >
        <div
            className="a b"
        />
    </rule>
</rule>,
]
`;

exports[`Creating Components with a Proxy for props from Fela rules should compose passThrough props 1`] = `
Array [
  "<style></style>",
  <FelaComponent
    color="red"
    onClick={[Function]}
>
    <FelaComponent
        _felaTheme={Object {}}
        color="red"
        onClick={[Function]}
    >
        <FelaComponent
            _felaRule={[Function]}
            color="red"
            onClick={[Function]}
            passThrough={
                Array [
                    "onClick",
                    "onClick",
                    "color",
                  ]
            }
        >
            <FelaComponent
                _felaRule={[Function]}
                _felaTheme={Object {}}
                color="red"
                onClick={[Function]}
                passThrough={
                    Array [
                        "onClick",
                        "onClick",
                        "color",
                      ]
                }
            >
                <div
                    className=""
                    color="red"
                    onClick={[Function]}
                />
            </FelaComponent>
        </FelaComponent>
    </FelaComponent>
</FelaComponent>,
]
`;

exports[`Creating Components with a Proxy for props from Fela rules should compose styles 1`] = `
Array [
  "<style>
    .a {
        color: red
    }

    .b {
        font-size: 16px
    }

    .c {
        line-height: 1.2
    }
</style>",
  <anotherRule>
    <anotherRule
        _felaTheme={Object {}}
    >
        <rule
            _felaRule={[Function]}
            passThrough={Array []}
        >
            <rule
                _felaRule={[Function]}
                _felaTheme={Object {}}
                passThrough={Array []}
            >
                <div
                    className="a b c"
                />
            </rule>
        </rule>
    </anotherRule>
</anotherRule>,
]
`;

exports[`Creating Components with a Proxy for props from Fela rules should not pass props used in rules to the element 1`] = `
Array [
  "<style>
    .a {
        color: true
    }

    .b {
        font-size: 16
    }
</style>",
  <rule
    color={true}
    data-foo={true}
    onClick={false}
>
    <rule
        _felaTheme={Object {}}
        color={true}
        data-foo={true}
        onClick={false}
    >
        <div
            className="a b"
            data-foo={true}
            onClick={false}
        />
    </rule>
</rule>,
]
`;

exports[`Creating Components with a Proxy for props from Fela rules should only use passed props to render Fela rules 1`] = `
Array [
  "<style>
    .a {
        color: black
    }

    .b {
        font-size: 16px
    }
</style>",
  <rule
    color="black"
    data-foo={true}
>
    <rule
        _felaTheme={Object {}}
        color="black"
        data-foo={true}
    >
        <div
            className="a b"
            color="black"
            data-foo={true}
        />
    </rule>
</rule>,
]
`;

exports[`Creating Components with a Proxy for props from Fela rules should only use the rule name as displayName 1`] = `
Array [
  "<style>
    .a {
        color: red
    }

    .b {
        font-size: 16
    }
</style>",
  <Button
    as="button"
>
    <Button
        _felaTheme={Object {}}
        as="button"
    >
        <button
            className="a b"
        />
    </Button>
</Button>,
]
`;

exports[`Creating Components with a Proxy for props from Fela rules should pass all props to the element 1`] = `
Array [
  "<style>
    .a {
        font-size: 16
    }
</style>",
  <rule
    data-foo={true}
    onClick={false}
>
    <rule
        _felaTheme={Object {}}
        data-foo={true}
        onClick={false}
    >
        <div
            className="a"
            data-foo={true}
            onClick={false}
        />
    </rule>
</rule>,
]
`;

exports[`Creating Components with a Proxy for props from Fela rules should pass props except innerRef 1`] = `
Array [
  "<style>
    .a {
        color: black
    }

    .b {
        font-size: 16px
    }
</style>",
  <rule
    color="black"
    innerRef={[Function]}
>
    <rule
        _felaTheme={Object {}}
        color="black"
        innerRef={[Function]}
    >
        <div
            className="a b"
        />
    </rule>
</rule>,
]
`;

exports[`Creating Components with a Proxy for props from Fela rules should pass props used in rules specified in passThroughProps to the element 1`] = `
Array [
  "<style>
    .a {
        color: true
    }

    .b {
        font-size: 16
    }
</style>",
  <rule
    color={true}
    data-foo={true}
    onClick={false}
>
    <rule
        _felaTheme={Object {}}
        color={true}
        data-foo={true}
        onClick={false}
    >
        <div
            className="a b"
            color={true}
            data-foo={true}
            onClick={false}
        />
    </rule>
</rule>,
]
`;

exports[`Creating Components with a Proxy for props from Fela rules should render fela rules depending on the passed props 1`] = `
Array [
  "<style>
    .a {
        color: black
    }

    .b {
        font-size: 16
    }
</style>",
  <rule
    color="black"
>
    <rule
        _felaTheme={Object {}}
        color="black"
    >
        <div
            className="a b"
        />
    </rule>
</rule>,
]
`;

exports[`Creating Components with a Proxy for props from Fela rules should use a dev-friendly className and the selectorPrefix 1`] = `
Array [
  "<style>
    .Fela-Button_div__abrv9k {
        font-size: 16
    }
</style>",
  <Button>
    <Button
        _felaTheme={Object {}}
    >
        <div
            className="Fela-Button_div__abrv9k"
        />
    </Button>
</Button>,
]
`;

exports[`Creating Components with a Proxy for props from Fela rules should use a dev-friendly className with monolithic renderer 1`] = `
Array [
  "<style>
    .Button_div__abrv9k {
        font-size: 16
    }
</style>",
  <Button>
    <Button
        _felaTheme={Object {}}
    >
        <div
            className="Button_div__abrv9k"
        />
    </Button>
</Button>,
]
`;

exports[`Creating Components with a Proxy for props from Fela rules should use the theme for static rendering by default 1`] = `
Array [
  "<style>
    .a {
        color: black
    }

    .b {
        font-size: 16
    }
</style>",
  <rule>
    <rule
        _felaTheme={
            Object {
                "color": "black",
              }
        }
    >
        <div
            className="a b"
        />
    </rule>
</rule>,
]
`;
